packages/prover/src/web3_proxy.ts

```diff
@@ -0,0 +1,122 @@
+import http from "node:http";
+import https from "node:https";
+import url from "node:url";
+import httpProxy from "http-proxy";
+import {NetworkName} from "@lodestar/config/networks";
+import {Logger} from "@lodestar/utils";
+import {LightNode} from "./interfaces.js";
+import {ProofProvider} from "./proof_provider/proof_provider.js";
+import {ELRequestPayload, ELResponse} from "./types.js";
+import {processAndVerifyRequest} from "./utils/execution.js";
+import {logRequest} from "./utils/logger.js";
+import {generateRPCResponseForPayload} from "./utils/json_rpc.js";
+import {fetchRequestPayload, fetchResponseBody} from "./utils/req_resp.js";
+
+export type VerifiedProxyOptions = {
+  network: NetworkName;
+  executionRpcUrl: string;
+  logger: Logger;
+  checkpoint?: string;
+} & ({mode: LightNode.Rest; urls: string[]} | {mode: LightNode.P2P; bootnodes: string[]});
+
+export function createVerifiedExecutionProxy(
+  opts: VerifiedProxyOptions
+): {server: http.Server; proofProvider: ProofProvider} {
+  const {executionRpcUrl: executionUrl, logger, network} = opts;
+  const controller = new AbortController();
+
+  const proofProvider = ProofProvider.init({
+    ...opts,
+    network: network,
+    signal: controller.signal,
+  });
+
+  const proxy = httpProxy.createProxy({
+    target: executionUrl,
+    ws: executionUrl.startsWith("ws"),
+    agent: https.globalAgent,
+    xfwd: true,
+    ignorePath: true,
+    changeOrigin: true,
+  });
+
+  let proxyServerListeningAddress: {host: string; port: number} | undefined;
+
+  function handler(payload: ELRequestPayload): Promise<ELResponse | undefined> {
+    return new Promise((resolve, reject) => {
+      if (!proxyServerListeningAddress) return reject(new Error("Proxy server not listening"));
+      const req = http.request(
+        {
+          method: "POST",
+          path: "/proxy",
+          port: proxyServerListeningAddress.port,
+          host: proxyServerListeningAddress.host,
+          signal: controller.signal,
+          headers: {
+            "Content-Type": "application/json",
+          },
+        },
+        (res) => {
+          fetchResponseBody(res).then(resolve).catch(reject);
+        }
+      );
+      req.write(JSON.stringify(payload));
+      req.end();
+    });
+  }
+
+  const proxyServer = http.createServer(function proxyRequestHandler(req, res) {
+    if (req.url === "/proxy") {
```

Comment:
would like to keep the url as `/prover` for better presentation

Reply:
The `/` is the actual prover url and `/proxy` is just forwarding endpoint. So this seems ok to me.